odata-protocol/11.4 Data Modification.md

```diff
@@ -654,49 +668,72 @@ operations.
+1. consists of an entity set followed by a key predicate
+2. consists of a singleton (which must be nullable if the request can be treated as an insert)
+3. ends with a containment navigation property,
+   followed by a key predicate if this is collection-valued
+   (and which must be nullable otherwise if the request can be treated as an insert)
+4. matches a navigation property path binding [#OData-CSDL#NavigationPropertyPathBinding]
 
-- containment navigation properties,
-- non-containment navigation properties with a navigation property binding, or
-- payloads including a context URL specifying the entity set or
-  contained collection of entities in which the new entity is to be created.
+or the payload includes a context URL.
 
 Upserts are not supported against entities whose keys' values are
-generated by the service. Services MUST fail an update request to a URL
+generated by the service. Services MUST fail an upsert request to a URL
 that would identify such an entity and the entity does not yet exist.
 
-Similarly, services MUST fail an update request to the URL of a [media entity](#RequestingtheMediaStreamofaMediaEntityusingvalue) that does not yet exist.
-However, a `PUT` request to the _media edit URL_ of a media entity does have Upsert
+Similarly, services MUST fail an upsert request to the URL of a [media entity](#RequestingtheMediaStreamofaMediaEntityusingvalue) that does not yet exist.
+However, a `PUT` request to the _media edit URL_ of a media entity does have upsert
 semantics, in that the media entity is [created](#CreateaMediaEntity)
 with the specified media stream if it does not already exist, otherwise the
 media stream of the existing media entity is
 [updated](#UpdateaMediaEntityStream).
 
-Singleton entities can be upserted if they are nullable. Services
-supporting this SHOULD advertise it by annotating the singleton with the
-term `Capabilities.UpdateRestrictions` (nested property `Upsertable`
-with value `true`) defined in [OData-VocCap](#ODataVocCap).
-
 A key property whose value is provided in the request URL SHOULD be omitted from the request body.
 If key properties are provided in the request URL and the request body with different values,
 services MUST either fail the request or ignore the value in the request body.
 
-To ensure that an update request is not treated as an insert, the client
-MAY specify an [`If-Match`](#HeaderIfMatch) header in the update
-request. The service MUST NOT treat an update request containing an
-`If-Match` header as an insert.
+If the URL of an upsert request falls into case 1, 2, or 3 given above and the request
+- contains an [`If-Match: *`](#HeaderIfMatch) header, the service MUST fail the request
+  if it would be treated as an insert
+- contains an `If-Match` header with an ETag as value, the service fails the request
+  instead of treating it as an insert because the [ETag](#UseofETagsforAvoidingUpdateConflicts)
+  does not match
```

Comment:
this leaves out the fact that the service can treat it as an update.
```suggestion
- contains an `If-Match` header with an ETag as value, the service treats the request
  as an update if the ETag matches the existing entity, otherwise fails the request
```